File: backend/api/tags.go
package api

import (
	"net/http"
	"sort"
	"strings"
	"time"

	"github.com/labstack/echo/v4"
	models "github.com/mannx/Bluebook/models"
	"github.com/rs/zerolog/log"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type tagData struct {
	models.TagList

	TagCount int
}

func getTagDataWithCount(db *gorm.DB) ([]tagData, error) {
	var tl []models.TagList

	res := db.Order("Tag").Find(&tl)
	if res.Error != nil {
		log.Error().Err(res.Error).Msg("Unable to retrieve tag list")
		return nil, res.Error
	}

	var td []tagData
	for _, i := range tl {
		// get the number of entries using this tag
		var data []models.TagData

		res = db.Where("TagID = ?", i.ID).Find(&data)
		if res.Error != nil {
			return nil, res.Error
		}

		td = append(td, tagData{
			TagList:  i,
			TagCount: len(data),
		})

	}

	return td, nil
}

// TagListViewHandler returns a list of all tags paired with their id
func TagListViewHandler(c echo.Context, db *gorm.DB) error {
	td, err := getTagDataWithCount(db)
	if err != nil {
		return LogAndReturnError(c, "Unable to retrieve tag counts", err)
	}

	return c.JSON(http.StatusOK, &td)
}

// TagDataViewHandler returns all the days that use the given tag. Provided by parameter ID
func TagDataViewHandler(c echo.Context, db *gorm.DB) error {
	type TagList struct {
		Day    models.DayData //data for day with the tag
		Date   string         // friendly string for the date
		Tags   []string       // list of all other tags associated with this day
		TagIDs []uint
	}

	var id uint

	err := echo.QueryParamsBinder(c).
		Uint("id", &id).
		BindError()
	if err != nil {
		return LogAndReturnError(c, "[TagDataViewHandler] Unable to bind to parameter: id", err)
	}

	var vd []models.TagData
	res := db.Find(&vd, "TagID = ?", id) // get all the days with this tag
	if res.Error != nil {
		return LogAndReturnError(c, "Unable to retrieve tag data (1)", res.Error)
	}

	var data []TagList

	for _, i := range vd {
		var d models.DayData

		err := db.Find(&d, "ID = ?", i.DayID) // get the day data
		if err.Error != nil {
			log.Warn().Err(err.Error).Msgf("Unable to get daydata [%v] from tagdata [%v]", i.DayID, i.ID)
			continue // skip if bad
		}

		//	DEBUG INFO
		//		if we have a tag for a day that no longer exists, it will be logged here
		//		currently this requires manual clean up of the db.
		//		ideally the data should be linked and deleted the day should delete the tag data also
		if err.RowsAffected == 0 {
			log.Info().Msgf("  DayID (%v) for tag (%v) not found in day_data", i.DayID, i.ID)
		}

		tstr, tids := GetTags(d.ID, db)
		tl := TagList{
			Day:    d,
			Date:   (time.Time(d.Date)).Format("Mon Jan _2 2006"),
			Tags:   tstr,
			TagIDs: tids,
		}

		data = append(data, tl)
	}

	// sort data by Day.Date
	sort.Slice(data, func(i, j int) bool {
		d1 := time.Time(data[i].Day.Date)
		d2 := time.Time(data[j].Day.Date)
		return d1.Before(d2)
	})

	return c.JSON(http.StatusOK, &data)

}

// TagUpdateViewHandler handles the POST'd data to update a day's tags
func TagUpdateViewHandler(c echo.Context, db *gorm.DB) error {
	type tagUpdate struct {
		Tag      string    `json:"Tag" query:"Tag"`
		LinkedID uint      `json:"LinkedID" query:"LinkedID"`
		Date     time.Time `json:"Date" query:"Date"`
	}

	var tu tagUpdate
	if err := c.Bind(&tu); err != nil {
		return LogAndReturnError(c, "Unable to bind from POST", err)
	}

	// are we updating tags on a linked day?
	if tu.LinkedID != 0 {
		// update tags linked to this day
		updateTags(tu.LinkedID, tu.Tag, db)

	} else {
		// attempt to generate an empty day and then try and link again
		id, err := genEmptyDay(db, tu.Date)
		if err != nil {
			return LogAndReturnError(c, "Unable to add tag to empty day", err)
		}

		updateTags(id, tu.Tag, db)
	}

	return nil
}

// generate the empty day, save it to the database and return its id
//
//	check to see if a day already exists and return its id instead of creating a new entry
//	if only a comment is present, no linked id will be sent and we end up here
func genEmptyDay(db *gorm.DB, date time.Time) (uint, error) {
	var data models.DayData
	res := db.Find(&data, "Date = ?", date)
	if res.Error != nil {
		return 0, res.Error
	}

	if res.RowsAffected != 0 {
		// found an entry, return its id
		return data.ID, nil
	}

	dd := models.DayData{
		Date: datatypes.Date(date),
	}

	res = db.Create(&dd)
	if res.Error != nil {
		return 0, res.Error
	}

	return dd.ID, nil
}

// parse and update the tags for a given day
func updateTags(id uint, tags string, db *gorm.DB) {

	// 0) Delete any tag data for the given day if currently present
	// 1) split tagStr into seperate tags
	//	a) remove any spaces and convert to lower case
	//	b) split string on each hash tag
	// 2) if found in the tag table, retrieve its id
	//	else add to the table and retrieve its new id
	// 3) add each tag id with the day id to the tag table

	// delete any tags currently linked to this day
	db.Delete(&models.TagData{}, "DayID = ?", id)

	tstr := strings.Split(tags, "#")
	for _, s := range tstr {
		if len(s) == 0 {
			continue // skip empty tags
		}

		// 2) remove any excess whitespace and check if in table
		str := strings.TrimSpace(s)

		var td models.TagList

		res := db.Find(&td, "Tag = ?", str)
		if res.Error != nil {
			log.Error().Err(res.Error).Msg("Unable to find tag")
			continue
		}

		if res.RowsAffected == 0 {
			// tag not found, add it and grab its id
			td.Tag = str
			db.Save(&td)
		}

		go func() {
			td := models.TagData{
				TagID: td.ID,
				DayID: id,
			}

			db.Save(&td)
		}()
	}
}

// TagCleanHandler removes all unused tags
func TagCleanHandler(c echo.Context, db *gorm.DB) error {
	// get list of all tags and their counts
	td, err := getTagDataWithCount(db)
	if err != nil {
		return LogAndReturnError(c, "Unable to retrieve tag data", err)
	}

	ids := make([]uint, 0)
	for _, i := range td {
		if i.TagCount == 0 {
			// add id to list
			ids = append(ids, i.ID)
		}
	}

	// remove all ids from the db
	db.Delete(&models.TagList{}, ids)

	return ReturnServerMessage(c, "Success", false)
}

File: backend/src/api/tags.rs
#![allow(non_snake_case)]
use diesel::prelude::*;
use diesel::SqliteConnection;
use log::{debug, error, info};
use serde::{Deserialize, Serialize};

use crate::api::DbError;
use crate::models::day_data::DayData;
use crate::models::tags::TagList;

// Tags is reused else where.  move to another mod with get_tags?
#[derive(Serialize, Deserialize)]
pub struct Tags {
    pub tag: String,
    pub id: i32,
}

#[derive(Serialize)]
pub struct TagInfo {
    id: i32,
    Tag: Option<String>,
    Count: i64,
}

impl TagInfo {
    fn new(tag: &TagList) -> Self {
        Self {
            id: tag.id,
            Tag: tag.Tag.clone(),
            Count: 0,
        }
    }
}

#[derive(Serialize)]
pub struct TagData {
    Day: DayData,
    Tags: Vec<Tags>,
}

pub fn get_tag_list(conn: &mut SqliteConnection) -> Result<Vec<TagInfo>, DbError> {
    use crate::schema::tag_list::dsl::*;

    let results = tag_list
        .order(Tag)
        .select(TagList::as_select())
        .load(conn)?;

    let mut tags = Vec::new();

    for tag in &results {
        let mut t = TagInfo::new(tag);
        t.Count = get_tag_count(conn, t.id)?;
        tags.push(t);
    }

    Ok(tags)
}

// return the number of days that have used this tag
fn get_tag_count(conn: &mut SqliteConnection, tag_id: i32) -> Result<i64, DbError> {
    use crate::schema::tag_data::dsl::*;

    let count = tag_data
        .filter(TagID.eq(tag_id))
        .count()
        .get_result::<i64>(conn)?;

    Ok(count)
}

pub fn get_tag_data(conn: &mut SqliteConnection, tag_id: i32) -> Result<Vec<TagData>, DbError> {
    use crate::schema::day_data::dsl::*;

    // get this list of days that have this tag associated with it
    let results: Vec<DayData> = day_data
        .filter(Tags.like(format!("%{tag_id}%")))
        .order(DayDate)
        .select(DayData::as_select())
        .load(conn)?;

    let mut output = Vec::new();

    // for each day we have retrieve the tags linked to it
    for day in results {
        let tags = get_tags(conn, &day)?;

        output.push(TagData {
            Day: day,
            Tags: tags,
        })
    }

    Ok(output)
}

// get all tags for a given day based on id
// this is used in serveral locations
pub fn get_tags(conn: &mut SqliteConnection, day: &DayData) -> Result<Vec<Tags>, DbError> {
    // retrieve all tags for this day
    let mut tags = Vec::new();

    // retrieve the list of all the tags for this day
    let tids = match &day.Tags {
        None => return Ok(tags),
        Some(tags) => {
            debug!("[get_tags] processing tags for [{tags}]");

            if tags.is_empty() {
                debug!("tag is empty [day: {}]. returning...", day.id);
                return Ok(Vec::new());
            }

            tags.split(' ')
                .map(|x| x.parse::<i32>().unwrap())
                .collect::<Vec<i32>>()
        }
    };

    for tag_id in tids {
        // get the tag text
        use crate::schema::tag_list::dsl::*;

        let tdata: TagList = tag_list.filter(id.eq(tag_id)).first::<TagList>(conn)?;
        tags.push(Tags {
            id: tdata.id,
            tag: tdata.Tag.unwrap_or_else(|| "".to_owned()),
        });
    }

    Ok(tags)
}

// migrate tags from the tag_data table into the DayData table
// once we are fully running and no longer need to migrate an old db
// this can be removed along with migrate_tags_do()
pub fn migrate_tags(conn: &mut SqliteConnection) -> Result<Vec<String>, DbError> {
    use crate::models::day_data::DayData;
    use crate::schema::day_data::dsl::*;
    use diesel::prelude::*;

    info!("Performing tag migration...");
    let mut messages = Vec::new();
    let results: Vec<DayData> = day_data.select(DayData::as_select()).load(conn)?;

    for r in results {
        info!("Migrating Tags for Date: {}", r.DayDate);
        match migrate_tags_do(conn, &r) {
            Ok(_) => {}
            Err(err) => {
                error!("Unable to update tags for day: {}", r.DayDate);
                error!("\t** {err}");

                messages.push(format!("Unable to update tags for day: {}", r.DayDate));
                messages.push(format!("\t**{err}"));
            }
        }
    }

    messages.push("Pruning unused tags...".to_owned());
    let mut pmsg = prune_tags(conn)?;
    messages.append(&mut pmsg);

    messages.push("Migration Success".to_owned());
    Ok(messages)
}

// migrate the tags for a given day
fn migrate_tags_do(conn: &mut SqliteConnection, data: &DayData) -> Result<(), DbError> {
    use crate::models::tags::TagData;
    use diesel::prelude::*;

    // retrieve any tags from tag_data associated with this id
    use crate::schema::tag_data::dsl::*;

    let tags: Vec<TagData> = tag_data
        .filter(DayID.eq(data.id))
        .select(TagData::as_select())
        .load(conn)?;

    let tlist: String = tags
        .iter()
        .map(|t| t.TagID.to_string())
        .collect::<Vec<String>>()
        .join(" ");

    if !tlist.is_empty() {
        // update the tags
        use crate::schema::day_data::dsl::*;

        info!("  Updating tag list...");
        diesel::update(day_data)
            .filter(id.eq(data.id))
            .set(Tags.eq(Some(tlist)))
            .execute(conn)?;
    }

    Ok(())
}

// prune off any tags that are not in use
// currently only done on tag migration, but should be linked to an
// api endpoint
fn prune_tags(conn: &mut SqliteConnection) -> Result<Vec<String>, DbError> {
    let tags = get_tag_list(conn)?;
    let mut msg = Vec::new();

    conn.transaction(|conn| {
        for t in tags {
            if t.Count == 0 {
                msg.push(format!(
                    "Pruning tag: {}",
                    t.Tag.unwrap_or("--NO TAG NAME--".to_owned())
                ));

                use crate::schema::tag_list::dsl::*;

                diesel::delete(tag_list.filter(id.eq(t.id))).execute(conn)?;
            }
        }
        diesel::result::QueryResult::Ok(())
    })?;

    Ok(msg)
}
